File: Java-Core/src/main/java/presto/expression/ThisExpression.java
package presto.expression;

import presto.error.PrestoError;
import presto.error.SyntaxError;
import presto.parser.Dialect;
import presto.runtime.Context;
import presto.type.IType;
import presto.utils.CodeWriter;
import presto.value.IValue;

public class ThisExpression implements IExpression {

	@Override
	public IType check(Context context) throws SyntaxError {
		if(context!=null && !(context instanceof Context.InstanceContext))
			context = context.getParentContext();
		if( context instanceof Context.InstanceContext)
			return ((Context.InstanceContext)context).getInstanceType();
		else
			throw new SyntaxError("Not in an instance context!");
	}
	
	@Override
	public IValue interpret(Context context) throws PrestoError {
		if(context!=null && !(context instanceof Context.InstanceContext))
			context = context.getParentContext();
		if( context instanceof Context.InstanceContext)
			return ((Context.InstanceContext)context).getInstance();
		else
			throw new SyntaxError("Not in an instance context!");
	}
	
	@Override
	public void toDialect(CodeWriter writer) {
		if(writer.getDialect()==Dialect.O)
			writer.append("this");
		else
			writer.append("self");
		
	}
}

File: Java-CodeServer/src/main/resources/js/prompto/expression/ThisExpression.js
var InstanceContext = require("../runtime/Context").InstanceContext;
var SyntaxError = require("../error/SyntaxError").SyntaxError;

function ThisExpression() {
    return this;
}

ThisExpression.prototype.check = function(context) {
    if (context != null && !(context instanceof InstanceContext))
        context = context.getParentContext ();
    if (context instanceof InstanceContext)
        return context.instanceType;
    else
        throw new SyntaxError ("Not in an instance context!");
};


ThisExpression.prototype.interpret = function(context) {
    if (context != null && !(context instanceof InstanceContext))
        context = context.getParentContext ();
    if (context instanceof InstanceContext)
        return context.instance;
    else
        throw new SyntaxError ("Not in an instance context!");
};

ThisExpression.prototype.toDialect = function(writer) {
    writer.toDialect(this);
};

ThisExpression.prototype.toEDialect = function(writer) {
    writer.append("self");
};

ThisExpression.prototype.toODialect = function(writer) {
    writer.append("this");
};

ThisExpression.prototype.toSDialect = function(writer) {
    writer.append("self");
};

exports.ThisExpression = ThisExpression;
